Library/Homebrew/formula_auditor.rb

```diff
@@ -332,6 +332,23 @@ def audit_deps
       end
 
       return unless @core_tap
+
+      # Prevent formulae from declaring Linux-only dependencies on GCC,
+      # or else its dependency tree grows out of control.
+      on_linux_deps = formula.to_hash_with_variations.dig("variations", :x86_64_linux, "dependencies")
+      if on_linux_deps&.include? "gcc"
```

Comment:
This needs to also check whether `gcc` is a dependency on macOS or not. Take the following case:

```ruby
depends_on "gcc"

on_linux do
  depends_on "gettext"
end
```

Then, `#to_hash_with_variations` will return this:

```ruby
{
  ...
  "dependencies" => ["GCC"],
  ...
  "variations" => {
    x86_64_linux: {
      "dependencies" => ["gcc", "gettext"],
    },
  },
}
```

`gcc` still shows up in the linux deps even though it's also a macOS dep. In this case, I don't think it should be included in this audit.

This could be even more complex if `gcc` is needed in only some macOS versions since `gcc` may or may not show up in the main dependency list, but maybe  but that may be too much of an edge case to worry about right now.